File: hmf-common/src/main/java/com/hartwig/hmftools/common/cider/IgTcrGeneFile.java
package com.hartwig.hmftools.common.cider;

import com.hartwig.hmftools.common.genome.refgenome.RefGenomeVersion;
import com.hartwig.hmftools.common.genome.region.Strand;
import com.hartwig.hmftools.common.region.BaseRegion;
import com.hartwig.hmftools.common.utils.file.DelimFileWriter;
import com.hartwig.hmftools.common.utils.file.DelimFileReader;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.*;
import java.util.ArrayList;
import java.util.List;

public class IgTcrGeneFile
{
    public enum Column
    {
        gene,
        allele,
        region,
        functionality,
        sequence,
        contig,
        posStart,
        posEnd,
        strand,
        anchorStart,
        anchorEnd,
        anchorSequence
    }

    private static final Logger sLogger = LogManager.getLogger(IgTcrGeneFile.class);

    public static List<IgTcrGene> read(RefGenomeVersion refGenomeVersion)
    {
        String resourcePath = "/refgenome/";
        if(refGenomeVersion.is37())
        {
            resourcePath += "igtcr_gene.37.tsv";
        }
        else if(refGenomeVersion.is38())
        {
            resourcePath += "igtcr_gene.38.tsv";
        }
        else
        {
            sLogger.error("unknown ref genome version: {}", refGenomeVersion);
            throw new IllegalArgumentException("unknown ref genome version: " + refGenomeVersion);
        }

        List<IgTcrGene> igTcrGenes = new ArrayList<>();

        InputStream tsvStream = IgTcrGeneFile.class.getResourceAsStream(resourcePath);
        if(tsvStream == null)
        {
            sLogger.error("unable to find resource file: {}", resourcePath);
            throw new RuntimeException("unable to find resource file: " + resourcePath);
        }

        try(DelimFileReader reader = new DelimFileReader(new BufferedReader(new InputStreamReader(tsvStream))))
        {
            for(DelimFileReader.Row record : reader)
            {
                String geneName = record.get(Column.gene);
                String allele = record.get(Column.allele);
                IgTcrRegion region = IgTcrRegion.valueOf(record.get(Column.region));
                IgTcrFunctionality functionality = IgTcrFunctionality.fromCode(record.get(Column.functionality));
                String sequence = record.get(Column.sequence);
                String anchorSequence = record.getOrNull(Column.anchorSequence);

                String contig = record.getOrNull(Column.contig);
                BaseRegion geneLocation = null;
                BaseRegion anchorLocation = null;
                Strand strand = null;

                if(contig != null)
                {
                    contig = refGenomeVersion.versionedChromosome(contig);

                    int posStart = record.getInt(Column.posStart);
                    int posEnd = record.getInt(Column.posEnd);

                    String strandStr = record.get(Column.strand);

                    if(strandStr.isEmpty())
                    {
                        throw new RuntimeException("chromosome exists but strand is missing");
                    }

                    strand = Strand.fromChar(strandStr.charAt(0));

                    if(posStart <= 0 || posEnd <= 0)
                    {
                        throw new RuntimeException("chromosome exists but pos start or pos end is invalid");
                    }

                    geneLocation = new BaseRegion(posStart, posEnd);

                    Integer anchorStart = record.getIntOrNull(Column.anchorStart);
                    Integer anchorEnd = record.getIntOrNull(Column.anchorEnd);

                    if(anchorStart != null && anchorEnd != null)
                    {
                        anchorLocation = new BaseRegion(anchorStart, anchorEnd);
                    }
                }

                igTcrGenes.add(new IgTcrGene(
                        geneName, allele, region, functionality, sequence, contig, geneLocation, strand, anchorSequence, anchorLocation));
            }
        }

        return igTcrGenes;
    }

    public static void write(String tsvPath, List<IgTcrGene> geneList)
    {
        DelimFileWriter.write(tsvPath, Column.values(), geneList,
                (gene, row) ->
                {
                    row.set(Column.gene, gene.geneName());
                    row.set(Column.allele, gene.allele());
                    row.set(Column.region, String.valueOf(gene.region()));
                    row.set(Column.functionality, gene.functionality().toCode());
                    row.set(Column.sequence, gene.sequence());
                    row.setOrNull(Column.contig, gene.contigName());
                    row.setOrNull(Column.posStart, gene.genePosition() != null ? gene.genePosition().start() : null);
                    row.setOrNull(Column.posEnd, gene.genePosition() != null ? gene.genePosition().end() : null);
                    row.setOrNull(Column.strand, gene.geneStrand() != null ? gene.geneStrand().asChar() : null);
                    row.setOrNull(Column.anchorStart, gene.anchorPosition() != null ? gene.anchorPosition().start() : null);
                    row.setOrNull(Column.anchorEnd, gene.anchorPosition() != null ? gene.anchorPosition().end() : null);
                    row.setOrNull(Column.anchorSequence, gene.anchorSequence());
                }
        );
    }
}

File: cider/src/main/java/com/hartwig/hmftools/cider/genes/IgTcrGeneFile.kt
package com.hartwig.hmftools.cider.genes

import com.hartwig.hmftools.cider.IgTcrFunctionality
import com.hartwig.hmftools.cider.IgTcrGene
import com.hartwig.hmftools.cider.IgTcrRegion
import com.hartwig.hmftools.common.codon.Codons
import com.hartwig.hmftools.common.genome.refgenome.RefGenomeVersion
import com.hartwig.hmftools.common.genome.region.Strand
import com.hartwig.hmftools.common.utils.file.FileWriterUtils
import org.apache.commons.csv.CSVFormat
import org.apache.commons.csv.CSVRecord
import org.apache.logging.log4j.LogManager
import java.io.BufferedReader
import java.io.InputStreamReader

object IgTcrGeneFile
{
    enum class Column
    {
        gene,
        allele,
        region,
        functionality,
        primaryAssembly,
        assemblyName,
        chromosome,
        posStart,
        posEnd,
        strand,
        anchorStart,
        anchorEnd,
        anchorSequence,
        anchorAA
    }

    private val sLogger = LogManager.getLogger(IgTcrGeneFile::class.java)

    fun read(refGenomeVersion: RefGenomeVersion) : List<IgTcrGene>
    {
        val resourcePath: String = if (refGenomeVersion.is37)
        {
            "igtcr_gene.37.tsv"
        }
        else if (refGenomeVersion.is38)
        {
            "igtcr_gene.38.tsv"
        }
        else
        {
            sLogger.error("unknown ref genome version: {}", refGenomeVersion)
            throw IllegalArgumentException("unknown ref genome version: $refGenomeVersion")
        }

        val igTcrGenes = ArrayList<IgTcrGene>()

        val tsvStream = javaClass.classLoader.getResourceAsStream(resourcePath)
        if (tsvStream == null)
        {
            sLogger.error("unable to find resource file: {}", resourcePath)
            throw RuntimeException("unable to find resource file: $resourcePath")
        }
        BufferedReader(InputStreamReader(tsvStream)).use { reader ->
            val format = CSVFormat.Builder.create()
                .setDelimiter('\t')
                .setRecordSeparator('\n')
                .setHeader().setSkipHeaderRecord(true) // use first line header as column names
                .build()
            val records: Iterable<CSVRecord> = format.parse(reader)
            for (record in records)
            {
                val geneName = record[Column.gene]
                val allele = record[Column.allele]
                val region = IgTcrRegion.valueOf(record[Column.region])
                val functionality = IgTcrFunctionality.fromCode(record[Column.functionality])
                val isPrimaryAssembly = record[Column.primaryAssembly].toBoolean()
                val assemblyName = if (isPrimaryAssembly) null else record[Column.assemblyName].intern()
                var anchorSequence: String? = record[Column.anchorSequence]
                if (anchorSequence!!.isEmpty())
                    anchorSequence = null
                var chromosome = record[Column.chromosome].intern()
                var genomicLocation: GenomicLocation? = null
                var anchorLocation: GenomicLocation? = null
                if (chromosome.isNotEmpty())
                {
                    chromosome = refGenomeVersion.versionedChromosome(chromosome)

                    val posStart = record[Column.posStart].toInt()
                    val posEnd = record[Column.posEnd].toInt()

                    val strandStr = record[Column.strand]

                    if (strandStr.isEmpty())
                    {
                        throw RuntimeException("chromosome exist but strand is missing")
                    }

                    val strand: Strand = Strand.fromChar(strandStr[0])

                    if (posStart <= 0 || posEnd <= 0)
                    {
                        throw RuntimeException("chromosome exist but pos start or pos end invalid")
                    }
                    genomicLocation = GenomicLocation(chromosome, posStart, posEnd, strand, assemblyName)

                    val anchorStart = record[Column.anchorStart]
                    val anchorEnd = record[Column.anchorEnd]

                    if (anchorStart.isNotEmpty() && anchorEnd.isNotEmpty())
                    {
                        anchorLocation = GenomicLocation(chromosome, anchorStart.toInt(), anchorEnd.toInt(), strand, assemblyName)
                    }
                }
                
                igTcrGenes.add(
                    IgTcrGene(geneName = geneName, allele = allele, region = region, functionality = functionality,
                    geneLocation = genomicLocation, anchorSequence = anchorSequence, anchorLocation = anchorLocation)
                )
            }
        }

        return igTcrGenes
    }

    fun write(tsvPath: String, geneList: List<IgTcrGene>)
    {
        val csvFormat = CSVFormat.Builder.create()
            .setDelimiter('\t').setRecordSeparator('\n')
            .setHeader(Column::class.java)
            .build()

        csvFormat.print(FileWriterUtils.createBufferedWriter(tsvPath)).use { csvPrinter ->
            for (gene: IgTcrGene in geneList)
            {
                for (col in Column.values())
                {
                    when (col)
                    {
                        Column.gene -> csvPrinter.print(gene.geneName)
                        Column.allele -> csvPrinter.print(gene.allele)
                        Column.region -> csvPrinter.print(gene.region)
                        Column.functionality -> csvPrinter.print(gene.functionality.toCode())
                        Column.primaryAssembly -> csvPrinter.print(gene.geneLocation?.inPrimaryAssembly)
                        Column.assemblyName -> csvPrinter.print(gene.geneLocation?.altAssemblyName)
                        Column.chromosome -> csvPrinter.print(gene.geneLocation?.chromosome)
                        Column.posStart -> csvPrinter.print(gene.geneLocation?.posStart)
                        Column.posEnd -> csvPrinter.print(gene.geneLocation?.posEnd)
                        Column.strand -> csvPrinter.print(gene.geneLocation?.strand?.asChar())
                        Column.anchorStart -> csvPrinter.print(gene.anchorLocation?.posStart)
                        Column.anchorEnd -> csvPrinter.print(gene.anchorLocation?.posEnd)
                        Column.anchorSequence -> csvPrinter.print(gene.anchorSequence)
                        Column.anchorAA -> csvPrinter.print(if (gene.anchorSequence != null) Codons.aminoAcidFromBases(gene.anchorSequence) else null)
                        //Column.sequence -> csvPrinter.print(gene.sequence)
                    }
                }
                csvPrinter.println()
            }
        }
    }
}